pubnub-core/src/data/channel/wildcard_spec.rs

```diff
@@ -0,0 +1,192 @@
+use crate::data::channel;
+use std::convert::TryFrom;
+use std::fmt::{self, Display};
+use std::str::FromStr;
+
+/// A wildcard Channel specifier.
+///
+/// This type represents a value suitable for use with [wildcard subscribe]
+/// feature.
+///
+/// Currently you can have up to three levels deep with your channel segment
+/// hierarchies, `a.b.c`, for example.
+///
+/// Developers MUST use .* to denote a wildcard subscription. Just having a * at
+/// the end of channel names will NOT get translated to .* and will result in a
+/// subscribe error.
+///
+/// [wildcard subscribe]: https://support.pubnub.com/support/solutions/folders/14000109563
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Default)]
+pub struct WildcardSpec(String);
+
+impl WildcardSpec {
+    // See https://support.pubnub.com/support/solutions/articles/14000043664-how-many-channel-segments-are-supported-with-wildcard-subscribe-
+    fn is_valid(s: &str) -> bool {
+        if s.starts_with('.') {
+            // Cannot start with the dot.
+            return false;
+        }
+
+        // A simple, manually implemented, state machine to check the validity.
+        {
+            let mut was_dot = false;
+            let mut was_asterisk = false;
+            let mut dots_count = 0;
+            for c in s.chars() {
+                if was_asterisk {
+                    // Asterisk must be last.
+                    return false;
+                }
+
+                was_asterisk = false;
+                if was_dot {
+                    if c == '*' {
+                        was_asterisk = true;
+                    }
+                } else if c == '*' {
+                    // Asterisk must be prepended by a dot, but this one wasn't.
+                    return false;
+                }
+                let is_dot = c == '.';
+                if is_dot {
+                    dots_count += 1;
+                    if dots_count > 2 {
+                        // There must be at most three segments.
+                        return false;
+                    }
+                }
+                was_dot = is_dot;
+            }
+            if was_dot {
+                // Dot must be followed by asterisk if it's at the end.
+                return false;
+            }
+        }
+
+        true
+    }
+
+    /// Create a new [`WildcardSpec`] skipping the validity check.
+    #[must_use]
+    pub fn from_string_unchecked(s: String) -> Self {
+        Self(s)
+    }
+}
+
+impl TryFrom<String> for WildcardSpec {
+    type Error = String;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        if !Self::is_valid(&value) {
+            return Err(value);
+        }
+        Ok(Self(value))
+    }
+}
+
+impl FromStr for WildcardSpec {
+    type Err = ();
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if !Self::is_valid(s) {
+            return Err(());
+        }
+        Ok(Self(s.to_owned()))
+    }
+}
```

Comment:
These `From` impls skip the `channel::Name` validation.

Reply:
Yes, but I think this is correct, as the constraints for the names are different. I.e. they're related, but I couldn't figure out how to match them.